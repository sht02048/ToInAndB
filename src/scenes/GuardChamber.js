import Renderer from "../graphics/Renderer";
import Guard from "../entities/enemies/Guard";
import Cow from "../entities/enemies/Cow";

class GuardChamber extends Renderer {
  #emperorWidth = 71;
  #isDone = true;

  constructor() {
    super();

    this.guard = new Guard({
      x: this.canvasWidth / 2 - this.#emperorWidth / 2,
      y: -100,
    });
    this.cowLeftList = [
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -150,
      }),
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -200,
      }),
    ];
    this.cowRightList = [
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -150,
      }),
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -200,
      }),
    ];
  }

  update() {
    this.guard.update();
    this.cowLeftList.forEach((cow, index) =>
      cow.update(
        this.guard.x - (index + 1) * 100,
        this.guard.y - (index + 1) * 50,
      ),
    );
    this.cowRightList.forEach((cow, index) =>
      cow.update(
        this.guard.x + (index + 1) * 100,
        this.guard.y - (index + 1) * 50,
      ),
    );
  }

  render() {
    this.guard.render();
    this.cowLeftList.forEach((cow) => cow.render());
    this.cowRightList.forEach((cow) => cow.render());
  }

  setTarget(player) {
    this.guard.setTargetList(player);
    this.cowLeftList.forEach((cow) => cow.setTargetList(player));
    this.cowRightList.forEach((cow) => cow.setTargetList(player));
  }

  setSceneTargetList() {
    const sceneTargetList = [this.guard];

    this.cowLeftList.forEach((cow) => sceneTargetList.push(cow));
    this.cowRightList.forEach((cow) => sceneTargetList.push(cow));

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    if (!this.guard.isDestroyed && !this.guard.isVanished) {
      this.#isDone = false;
    }

    this.cowLeftList.forEach((cow) => {
      if (!cow.isDestroyed && !cow.isVanished) {
        this.#isDone = false;
      }
    });

    this.cowRightList.forEach((cow) => {
      if (!cow.isDestroyed && !cow.isVanished) {
        this.#isDone = false;
      }
    });

    return this.#isDone;
  }
}

export default GuardChamber;
