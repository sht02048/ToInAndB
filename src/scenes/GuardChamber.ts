import Renderer from "../graphics/Renderer";

import Player from "../entities/Player";
import Guard from "../entities/enemies/Guard";
import Cow from "../entities/enemies/Cow";

class GuardChamber extends Renderer {
  #emperorWidth = 71;
  #isDone = true;

  private guard: Guard;
  private cowLeftList: Cow[];
  private cowRightList: Cow[];
  private shouldBeDisplayed: boolean;
  private hasStarted: boolean;

  constructor() {
    super();

    this.guard = new Guard({
      x: this.canvasWidth / 2 - this.#emperorWidth / 2,
      y: -100,
    });
    this.cowLeftList = [
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -150,
      }),
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -200,
      }),
    ];
    this.cowRightList = [
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -150,
      }),
      new Cow({
        x: this.canvasWidth / 2 - this.#emperorWidth / 2,
        y: -200,
      }),
    ];

    this.shouldBeDisplayed = false;
    this.hasStarted = false;
  }

  update() {
    this.guard.update();
    this.cowLeftList.forEach((cow, index) =>
      cow.update(
        this.guard.x - (index + 1) * 100,
        this.guard.y - (index + 1) * 50,
      ),
    );
    this.cowRightList.forEach((cow, index) =>
      cow.update(
        this.guard.x + (index + 1) * 100,
        this.guard.y - (index + 1) * 50,
      ),
    );
  }

  render() {
    this.guard.render();
    this.cowLeftList.forEach((cow) => cow.render());
    this.cowRightList.forEach((cow) => cow.render());
  }

  setTarget(player: Player) {
    this.guard.setTargetList(player);
    this.cowLeftList.forEach((cow) => cow.setTargetList(player));
    this.cowRightList.forEach((cow) => cow.setTargetList(player));
  }

  setSceneTargetList() {
    const sceneTargetList = [this.guard, ...this.cowLeftList, ...this.cowRightList];

    return sceneTargetList;
  }

  checkSceneStatus() {
    this.#isDone = true;

    if (!this.guard.isDestroyed && !this.guard.isVanished) {
      this.#isDone = false;
    }

    this.cowLeftList.forEach((cow) => {
      if (!cow.isDestroyed && !cow.isVanished) {
        this.#isDone = false;
      }
    });

    this.cowRightList.forEach((cow) => {
      if (!cow.isDestroyed && !cow.isVanished) {
        this.#isDone = false;
      }
    });

    return this.#isDone;
  }
}

export default GuardChamber;
