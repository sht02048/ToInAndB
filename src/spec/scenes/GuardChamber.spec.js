import { beforeEach, describe, expect } from "vitest";
import GuardChamber from "../../scenes/GuardChamber";

describe("GuardChamber", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let guardChamber;

  beforeEach(() => {
    guardChamber = new GuardChamber();
  });

  it("should have proper functions", () => {
    expect(guardChamber.update).to.be.a("function");
    expect(guardChamber.render).to.be.a("function");
    expect(guardChamber.setTarget).to.be.a("function");
    expect(guardChamber.setSceneTargetList).to.be.a("function");
    expect(guardChamber.checkSceneStatus).to.be.a("function");
  });

  it("should render enemies", () => {
    guardChamber.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should render enemies", () => {
    guardChamber.update();
    guardChamber.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should check if the scene is done", () => {
    guardChamber.guard.isDestroyed = true;
    guardChamber.cowLeftList.forEach((cow) => (cow.isDestroyed = true));
    guardChamber.cowRightList.forEach((cow) => (cow.isDestroyed = true));

    expect(guardChamber.checkSceneStatus()).to.equal(true);
  });
});
