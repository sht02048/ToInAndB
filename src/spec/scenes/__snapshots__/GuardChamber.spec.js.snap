// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GuardChamber > should render enemies 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/emperor.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -150,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -150,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`GuardChamber > should render enemies 2`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -100,
      "img": <img
        src="/src/assets/sprites/enemies/emperor.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -150,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -150,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -200,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 114.5,
      "dy": -99,
      "img": <img
        src="/src/assets/sprites/enemies/emperor.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 14.5,
      "dy": -149,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 214.5,
      "dy": -149,
      "img": <img
        src="/src/assets/sprites/enemies/cow.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
